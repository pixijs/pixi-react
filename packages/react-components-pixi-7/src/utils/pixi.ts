import { invariant } from '@pixi/react-invariant';
import isNil from 'lodash.isnil';
import type { PixiReactContainer, PointCoords, PointLike } from '../types';
import type { IPoint } from '@pixi/core';

/**
 * Parse PIXI point value to array of coordinates
 *
 * @param {*} value
 * @returns {Array}
 */
export function parsePoint(value: PointLike | string): PointCoords
{
    let arr: (number | string)[] = [];

    if (typeof value === 'undefined')
    {
        return arr as PointCoords;
    }
    else if (typeof value === 'string')
    {
        arr = value.split(',');
    }
    else if (typeof value === 'number')
    {
        arr = [value];
    }
    else if (Array.isArray(value))
    {
        arr = [...value];
    }
    else if (value !== null && typeof value === 'object')
    {
        const x = (value && value?.x) || 0;
        const y = (value && value?.y) || 0;

        arr = [x, y];
    }
    else
    {
        return arr as PointCoords;
    }

    const isNumberLike = (p: any): boolean => !isNil(p) && !isNaN(p as number);

    return arr.filter(isNumberLike).map(Number) as PointCoords;
}

/**
 * Check if two points are equal
 *
 * @param {*} oldValue
 * @param {*} newValue
 * @returns {boolean}
 */
export function pointsAreEqual(oldValue: PointCoords, newValue: PointCoords)
{
    if (oldValue.length !== newValue.length)
    {
        return false;
    }

    for (let i = 0; i < oldValue.length; i++)
    {
        if (oldValue[i] !== newValue[i])
        {
            return false;
        }
    }

    return true;
}

/**
 * Determine value is type of Point or ObservablePoint
 * See https://github.com/michalochman/react-pixi-fiber/blob/a4dbddbef0ffbf0f563c64d30766ea28222a51ea/src/utils.js#L48
 *
 * @param {*} value
 * @returns {boolean}
 */
export function isPointType(value: any): value is IPoint
{
    // avoid instanceof check for perf
    return (
        !isNil(value)
        && value.x !== undefined
        && value.y !== undefined
        && value.copyFrom !== undefined
        && value.copyTo !== undefined
        && value.equals !== undefined
        && value.set !== undefined
    );
}

// TODO: can we pull this from FederatedEventEmitterTypes
/**
 * Event handlers
 *
 * @type {string[]}
 */
export const eventHandlers = [
    'click',
    'mousedown',
    'mousemove',
    'mouseout',
    'mouseover',
    'mouseup',
    'mouseupoutside',
    'pointercancel',
    'pointerout',
    'pointerover',
    'pointertap',
    'pointerdown',
    'pointerup',
    'pointerupoutside',
    'pointermove',
    'rightclick',
    'rightdown',
    'rightup',
    'rightupoutside',
    'tap',
    'touchcancel',
    'touchend',
    'touchendoutside',
    'touchmove',
    'touchstart',
];

/**
 * Set value on a PIXI.DisplayObject
 * See https://github.com/Izzimach/react-pixi/blob/a25196251a13ed9bb116a8576d93e9fceac2a14c/src/ReactPIXI.js#L114
 *
 * @param {PIXI.DisplayObject} instance
 * @param {string} prop
 * @param {*} value
 */
export function setValue(instance: PixiReactContainer, prop: string, value: any)
{
    type InstanceProperty = keyof typeof instance;

    const currentValue = instance[prop as InstanceProperty];
    const currentValueIsPointType = isPointType(currentValue);

    if (currentValueIsPointType && isPointType(value))
    {
        // copy value
        currentValue.copyFrom(value);
    }
    else if (currentValueIsPointType)
    {
        // parse value if a non-Point type is being assigned to a Point type
        const coordinates = parsePoint(value);

        invariant(
            typeof coordinates !== 'undefined' && coordinates.length > 0 && coordinates.length < 3,
            'The property `%s` is a `PIXI.Point` or `PIXI.ObservablePoint` and must be set to a comma-separated string of '
                + 'either 1 or 2 coordinates, a 1 or 2 element array containing coordinates, or a PIXI Point/ObservablePoint. '
                + 'If only one coordinate is given then X and Y will be set to the provided value. Received: `%s` of type `%s`.',
            prop,
            JSON.stringify(value),
            typeof value,
        );

        const [x, y] = coordinates;

        currentValue.set(x, y);
    }
    else
    {
        // just hard assign value
        // @ts-ignore - ignore readonly property value errors
        instance[prop as InstanceProperty] = value;
    }
}
