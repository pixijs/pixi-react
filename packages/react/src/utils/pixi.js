import { Point, ObservablePoint } from '@pixi/math';
import invariant from '../utils/invariant';
import isNil from 'lodash.isnil';

/**
 * Parse PIXI point value to array of coordinates
 *
 * @param {*} value
 * @returns {Array}
 */
export function parsePoint(value)
{
    let arr = [];

    if (typeof value === 'undefined')
    {
        return arr;
    }
    else if (typeof value === 'string')
    {
        arr = value.split(',');
    }
    else if (typeof value === 'number')
    {
        arr = [value];
    }
    else if (Array.isArray(value))
    {
        arr = [...value];
    }
    else if (value !== null && typeof value === 'object')
    {
        const x = (value && value?.x) || 0;
        const y = (value && value?.y) || 0;

        arr = [x, y];
    }
    else
    {
        return arr;
    }

    return arr.filter((p) => !isNil(p) && !isNaN(p)).map(Number);
}

/**
 * Check if two points are equal
 *
 * @param {*} oldValue
 * @param {*} newValue
 * @returns {boolean}
 */
export function pointsAreEqual(oldValue, newValue)
{
    if (oldValue.length !== newValue.length)
    {
        return false;
    }

    for (let i = 0; i < oldValue.length; i++)
    {
        if (oldValue[i] !== newValue[i])
        {
            return false;
        }
    }

    return true;
}

/**
 * Determine value is type of Point or ObservablePoint
 * See https://github.com/michalochman/react-pixi-fiber/blob/a4dbddbef0ffbf0f563c64d30766ea28222a51ea/src/utils.js#L48
 *
 * @param {*} value
 * @returns {boolean}
 */
export function isPointType(value)
{
    return value instanceof Point || value instanceof ObservablePoint;
}

/**
 * Event handlers
 *
 * @type {string[]}
 */
export const eventHandlers = [
    'click',
    'mousedown',
    'mousemove',
    'mouseout',
    'mouseover',
    'mouseup',
    'mouseupoutside',
    'tap',
    'touchstart',
    'touchmove',
    'touchend',
    'touchendoutside',
    'pointercancel',
    'pointerout',
    'pointerover',
    'pointertap',
    'pointerdown',
    'pointerup',
    'pointerupoutside',
    'pointermove',
    'rightclick',
    'rightdown',
    'rightup',
    'rightupoutside',
    'touchcancel',
    'touchendoutside',
    'touchmove',
    'touchstart',
];

/**
 * Set value on a PIXI.DisplayObject
 * See https://github.com/Izzimach/react-pixi/blob/a25196251a13ed9bb116a8576d93e9fceac2a14c/src/ReactPIXI.js#L114
 *
 * @param {PIXI.DisplayObject} instance
 * @param {string} prop
 * @param {*} value
 */
export function setValue(instance, prop, value)
{
    if (isPointType(instance[prop]) && isPointType(value))
    {
    // copy value
        instance[prop].copyFrom(value);
    }
    else if (isPointType(instance[prop]))
    {
    // parse value if a non-Point type is being assigned to a Point type
        const coordinates = parsePoint(value);

        invariant(
            typeof coordinates !== 'undefined' && coordinates.length > 0 && coordinates.length < 3,
            'The property `%s` is a `PIXI.Point` or `PIXI.ObservablePoint` and must be set to a comma-separated string of '
        + 'either 1 or 2 coordinates, a 1 or 2 element array containing coordinates, or a PIXI Point/ObservablePoint. '
        + 'If only one coordinate is given then X and Y will be set to the provided value. Received: `%s` of type `%s`.',
            prop,
            JSON.stringify(value),
            typeof value
        );

        instance[prop].set(coordinates.shift(), coordinates.shift());
    }
    else
    {
    // just hard assign value
        instance[prop] = value;
    }
}
