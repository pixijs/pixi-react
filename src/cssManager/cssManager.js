import {
  getToBeTweenedProps,
  convertCSSPropsToPIXI,
  getContexts,
  transitionPropertyNames,
  animationPropertyNames,
  textStylePropertiesNames,
} from '../utils/css'
import { convertAnimationToTween, convertTransitionToTween } from './tweens'
import cssSource from './cssSource'
import { unassign, pick, isEmpty, splitAndTrim } from '../helpers'

const filterTweenPropsAndParams = (props, toRemove, action) => {
  const filter = {
    in: pick,
    out: unassign,
  }
  const transitionProps = props.transitionProperty
  const propsHandledByTransition = Object.keys(toRemove)

  return filter[action](props, [...propsHandledByTransition, ...transitionPropertyNames, ...animationPropertyNames])
}

const getCSSBasedProps = cssRule => {
  const props = {
    style: {},
  }

  Object.keys(cssRule).forEach(rule => {
    if (textStylePropertiesNames.includes(rule)) {
      props.style[rule] = cssRule[rule]
    } else {
      props[rule] = cssRule[rule]
    }
  })
  return props
}

const extendPIXIWithCSSProps = (cssProps, props = {}) => {
  const elProps = convertCSSPropsToPIXI(cssProps)
  const importantElProps = convertCSSPropsToPIXI(cssProps, true)
  return {
    ...elProps,
    ...props,
    ...importantElProps,
    style: {
      ...elProps.style,
      ...props.style,
      ...importantElProps.style,
    },
  }
}

const getInitialElementProp = (element, cssBasedProp) => {
  let value
  if (isEmpty(element[cssBasedProp])) {
    const context = getContexts()[cssBasedProp]
    if (context) {
      const axis =
        cssBasedProp
          .charAt(cssBasedProp.length - 1)
          .toLowerCase()
          .indexOf('x') !== -1
          ? 'x'
          : 'y'
      value = element[context][axis]
    } else {
      value = element[cssBasedProp]
    }
  } else {
    value = element[cssBasedProp]
  }
  return value
}

export default (element, type, applyProps) => {
  const initialCSSValues = {}
  let allProps
  let currentClassName
  let currentProps = {}
  let transitionProps = {}
  element.getCSSBasedProps = () => {
    return extendPIXIWithCSSProps(getCSSBasedProps(cssSource.getCSSRuleByName(currentClassName)))
  }
  const tweenComponent = (animationProps, animationParams) => {
    const { transitionProperty, animationName } = animationParams
    const arrTransitionProps = Object.keys(transitionProps)
    if (animationName) {
      const animationByName = cssSource.getCSSAnimationByName(animationName)
      if (animationByName) {
        convertAnimationToTween(element, animationByName, animationParams)
      }
    }
    if (arrTransitionProps.length > 0) {
      convertTransitionToTween(element, pick(animationProps, arrTransitionProps), animationParams)
    }

    if (transitionProperty) {
      transitionProperty.split(',').forEach(p => {
        transitionProps[p.trim()] = true
      })
    }
  }

  const setCSSProps = (className, isUpdate, props) => {
    let animationProps
    let animationParams
    if (className) {
      if (!isUpdate && currentProps.styled) {
        const styledComponentCSSSelector = className.split(' ')
        currentClassName = styledComponentCSSSelector[styledComponentCSSSelector.length - 1]
        cssSource.addStyleSheet(document.querySelectorAll('[data-styled]')[0].innerHTML, currentClassName)
      } else {
        currentClassName = className
      }
    } else {
      transitionProps = {}
    }

    const cssBasedProps = getCSSBasedProps(cssSource.getCSSRuleByName(className))
    Object.keys(cssBasedProps).forEach(cssBasedProp => {
      if (isEmpty(initialCSSValues[cssBasedProp])) {
        initialCSSValues[cssBasedProp] = getInitialElementProp(element, cssBasedProp)
      }
    })

    allProps = {
      ...initialCSSValues,
      ...extendPIXIWithCSSProps(cssBasedProps, props),
    }

    animationProps = {
      ...initialCSSValues,
      ...filterTweenPropsAndParams(allProps, transitionProps, 'in'),
    }

    animationParams = getToBeTweenedProps(animationProps)
    const { transitionProperty } = animationParams
    if (!transitionProperty) {
      transitionProps = {}
    }

    const propsAndParamsWithoutTweened = filterTweenPropsAndParams(allProps, transitionProps, 'out')
    tweenComponent(animationProps, animationParams)

    return propsAndParamsWithoutTweened
  }

  const updateComponent = () => {
    if (!cssSource.hasMediaQuery(`.${currentClassName}`)) return
    setCSSProps(currentClassName, true)

    if (applyProps) {
      applyProps(element, currentProps, filterTweenPropsAndParams(allProps, transitionProps, 'out'))
    }
  }

  const setProps = props => {
    currentProps = props
  }

  cssSource.addMediaQueryListener(updateComponent)

  element.on('removed', () => {
    //clean up when element is removed by the stage
    delete element.cssManager
    cssSource.removeMediaQueryListener(updateComponent)
  })

  return {
    setCSSProps,
    setProps,
  }
}
