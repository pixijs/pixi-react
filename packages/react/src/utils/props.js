import { Texture } from '@pixi/core';
import { DisplayObject } from '@pixi/display';
import { eventHandlers, setValue } from './pixi';
import invariant from '../utils/invariant';
import { not, hasKey } from '../helpers';

export const CHILDREN = 'children';
/**
 * Reserved props
 *
 * @type {Object}
 */
export const PROPS_RESERVED = {
    [CHILDREN]: true,
    parent: true,
    worldAlpha: true,
    worldTransform: true,
    worldVisible: true,
};

/**
 * Default display object props
 * See https://github.com/michalochman/react-pixi-fiber/blob/a4dbddbef0ffbf0f563c64d30766ea28222a51ea/src/props.js#L9
 *
 * @type {Object}
 */
export const PROPS_DISPLAY_OBJECT = {
    alpha: 1,
    buttonMode: false,
    cacheAsBitmap: false,
    cursor: null,
    filterArea: null,
    filters: null,
    hitArea: null,
    interactive: false,
    mask: null,
    pivot: 0,
    position: 0,
    renderable: true,
    rotation: 0,
    scale: 1,
    skew: 0,
    transform: null,
    visible: true,
    x: 0,
    y: 0,
};

/**
 * Helper util for fetching the texture from props
 * Can be either texture or image
 *
 * @param {string} elementType
 * @param {PIXI.Container} root
 * @param {object} props
 * @returns {PIXI.Texture|null}
 */
export const getTextureFromProps = (elementType, root, props = {}) =>
{
    const emitChange = (texture) =>
        requestAnimationFrame(() =>
        {
            texture?.__reactpixi?.root?.emit(`__REACT_PIXI_REQUEST_RENDER__`);
        });

    // eslint-disable-next-line consistent-return
    const check = (inType, validator) =>
    {
        if (props.hasOwnProperty(inType))
        {
            const valid
        = validator.typeofs.some((t) => typeof props[inType] === t)
        || validator.instanceofs.some((i) => props[inType] instanceof i);

            invariant(valid, `${elementType} ${inType} prop is invalid`);

            return props[inType];
        }
    };

    if (props.texture)
    {
        invariant(props.texture instanceof Texture, `${elementType} texture needs to be typeof \`Texture\``);

        return props.texture;
    }
    const result
      = check('image', { typeofs: ['string'], instanceofs: [HTMLImageElement] })
      || check('video', { typeofs: ['string'], instanceofs: [HTMLVideoElement] })
      || check('source', {
          typeofs: ['string', 'number'],
          instanceofs: [HTMLImageElement, HTMLVideoElement, HTMLCanvasElement, Texture],
      });

    invariant(!!result, `${elementType} could not get texture from props`);

    const texture = Texture.from(result);

    texture.__reactpixi = { root };
    texture.once('update', emitChange);
    texture.once('loaded', emitChange);

    if (texture.valid)
    {
        emitChange(texture);
    }

    return texture;
};

const filterProps = not(hasKey([...Object.keys(PROPS_RESERVED), ...eventHandlers]));

/**
 * Apply default props on Display Object instance (which are all components)
 *
 * @param {PIXI.DisplayObject} instance
 * @param {Object} oldProps
 * @param {Object} newProps
 */
export function applyDefaultProps(instance, oldProps, newProps)
{
    let changed = false;

    invariant(
        DisplayObject.prototype.isPrototypeOf(instance),
        'instance needs to be typeof `DisplayObject`, got `%s`',
        typeof instance
    );

    // update event handlers
    if (!newProps.ignoreEvents)
    {
        const hasRemoveListener = typeof instance.removeListener === 'function';
        const hasOn = typeof instance.on === 'function';

        for (let i = 0; i < eventHandlers.length; i++)
        {
            const evt = eventHandlers[i];

            if (oldProps[evt] !== newProps[evt])
            {
                changed = true;
                if (typeof oldProps[evt] === 'function' && hasRemoveListener)
                {
                    instance.removeListener(evt, oldProps[evt]);
                }
                if (typeof newProps[evt] === 'function' && hasOn)
                {
                    instance.on(evt, newProps[evt]);
                }
            }
        }
    }

    const newPropKeys = Object.keys(newProps || {});

    // hard overwrite all props? can speed up performance
    if (newProps.overwriteProps)
    {
        for (let i = 0; i < newPropKeys.length; i++)
        {
            const p = newPropKeys[i];

            if (oldProps[p] !== newProps[p])
            {
                changed = true;
                setValue(instance, p, newProps[p]);
            }
        }

        // TODO: previously this returned nothing ie. falsy, should we explicitly return false here instead or was it always
        // a subtle bug, test the intention
        return changed;
    }

    const props = newPropKeys.filter(filterProps);

    for (let i = 0; i < props.length; i++)
    {
        const prop = props[i];
        const value = newProps[prop];

        if (newProps[prop] !== oldProps[prop])
        {
            changed = true;
        }

        if (value !== undefined)
        {
            // set value if defined
            setValue(instance, prop, value);
        }
        else if (prop in PROPS_DISPLAY_OBJECT)
        {
            // is a default value, use that
            console.warn(`setting default value: ${prop}, from: ${instance[prop]} to: ${value} for`, instance);
            changed = true;
            setValue(instance, prop, PROPS_DISPLAY_OBJECT[prop]);
        }
        else
        {
            console.warn(`ignoring prop: ${prop}, from ${instance[prop]} to ${value} for`, instance);
        }
    }

    return changed;
}
