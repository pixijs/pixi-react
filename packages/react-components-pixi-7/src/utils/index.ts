import * as props from './props';
import * as pixi from './pixi';

export { props, pixi };
