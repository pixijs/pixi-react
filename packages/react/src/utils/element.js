import TYPES from '@pixi/react-tag-types';

import invariant from '../utils/invariant';
import { applyDefaultProps } from './props';
import * as components from '../components';

const ELEMENTS = Object.keys(TYPES).reduce((elements, type) => ({ ...elements, [type]: components[type] }), {});

/**
 * Inject types
 *
 * @type {Object}
 */
export const TYPES_INJECTED = {};

/**
 * Create an element based on tag type
 * Similar to react-dom's `React.createElement()`
 *
 * @param {string} type Element type
 * @param {Object} props Component props
 * @param {Object} root Root instance
 */
export function createElement(type, props = {}, root = null)
{
    const fn = ELEMENTS[type];

    let instance;
    let applyProps;

    if (typeof fn === 'function')
    {
        instance = fn(root, props);
    }

    if (!instance)
    {
    // not found, is there any injected custom component?
        const injected = TYPES_INJECTED[type];

        if (injected)
        {
            instance = injected.create(props);
            instance.didMount = injected.didMount ? injected.didMount.bind(instance) : undefined;
            instance.willUnmount = injected.willUnmount ? injected.willUnmount.bind(instance) : undefined;
            instance.applyProps = injected.applyProps ? injected.applyProps.bind(instance) : undefined;
            instance.config = injected.config;
        }
    }

    // apply initial props!
    if (instance)
    {
        applyProps = typeof instance?.applyProps === 'function' ? instance.applyProps : applyDefaultProps;
        applyProps(instance, {}, props);

        instance.__reactpixi = {
            root,
        };
    }

    return instance;
}

/**
 * Create Component
 *
 * @param {string} type
 * @param {Object} lifecycle methods
 */
export function PixiComponent(type, lifecycle)
{
    invariant(!!type, 'Expect type to be defined, got `%s`', type);
    invariant(!TYPES[type], 'Component `%s` could not be created, already exists in default components.', type);

    TYPES_INJECTED[type] = lifecycle;

    return type;
}
