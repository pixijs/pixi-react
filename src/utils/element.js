import invariant from 'fbjs/lib/invariant'
import idx from 'idx'
import { applyDefaultProps } from './props'
import cssManager from '../cssManager/cssManager'
import * as components from '../components'

/**
 * Available tag types
 *
 * @type {Object}
 */
export const TYPES = {
  BitmapText: 'BitmapText',
  Container: 'Container',
  Graphics: 'Graphics',
  NineSlicePlane: 'NineSlicePlane',
  ParticleContainer: 'ParticleContainer',
  Sprite: 'Sprite',
  AnimatedSprite: 'AnimatedSprite',
  Text: 'Text',
  TilingSprite: 'TilingSprite',
  SimpleMesh: 'SimpleMesh',
  SimpleRope: 'SimpleRope',
}

const ELEMENTS = Object.keys(TYPES).reduce((elements, type) => ({ ...elements, [type]: components[type] }), {})

/**
 * Inject types
 *
 * @type {Object}
 */
export const TYPES_INJECTED = {}

/**
 * Create an element based on tag type
 * Similar to react-dom's `React.createElement()`
 *
 * @param {string} type Element type
 * @param {Object} props Component props
 * @param {Object} root Root instance
 * @returns {PIXI.*|undefined}
 */
export function createElement(type, { className, ...props } = {}, root = null) {
  const fn = ELEMENTS[type]
  let instance
  let applyProps
  let cssProps
  let cssMan
  if (typeof fn === 'function') {
    instance = fn(root, props)
  }

  if (!instance) {
    // not found, is there any injected custom component?
    const injected = TYPES_INJECTED[type]
    if (injected) {
      instance = injected.create(props)
      instance.didMount = injected.didMount ? injected.didMount.bind(instance) : undefined
      instance.willUnmount = injected.willUnmount ? injected.willUnmount.bind(instance) : undefined
      instance.applyProps = injected.applyProps ? injected.applyProps.bind(instance) : undefined
    }
  }

  // apply initial props!
  if (instance) {
    if (className) {
      cssMan = cssManager(instance, type, applyDefaultProps)
      instance.cssManager = cssMan
      cssMan.setProps(props)
      cssProps = cssMan.setCSSProps(className, null, props)
    }
    applyProps = idx(instance, _ => _.applyProps)
    if (typeof applyProps !== 'function') {
      applyProps = applyDefaultProps
    }

    applyProps(instance, {}, className ? cssProps : props)
  }

  return instance
}

/**
 * Create Component
 *
 * @param {string} type
 * @param {Object} lifecycle methods
 */
export function PixiComponent(type, lifecycle) {
  invariant(!!type, 'Expect type to be defined, got `%s`', type)
  invariant(!TYPES[type], 'Component `%s` could not be created, already exists in default components.', type)

  TYPES_INJECTED[type] = lifecycle

  return type
}
