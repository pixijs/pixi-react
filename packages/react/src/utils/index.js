import * as element from './element';
import * as props from './props';
import * as pixi from './pixi';

export { element, props, pixi };
