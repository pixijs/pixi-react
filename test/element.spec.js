import * as PIXI from 'pixi.js'
import { createElement, TYPES, TYPES_INJECTED, PixiComponent } from '../src/utils/element'
import cssManager from '../src/cssManager/cssManager'
import * as propsUtils from '../src/utils/props';
import { emptyTexture } from './__fixtures__/textures'
import { desyrel } from './__fixtures__/bitmapfonts'
import parseBitmapFont from './__utils__/parseBitmapFont'

jest.mock('../src/cssManager/cssManager');


cssManager.mockImplementation(() => {
  return {
    setCSSProps: jest.fn().mockReturnValue({}),
    getCSSProps: jest.fn(),
    setProps: jest.fn(),
  }
});


parseBitmapFont(desyrel)

describe('createElement', () => {
  test('types', () => {
    expect(TYPES).toMatchSnapshot()
  })

  test('create Container', () => {
    const element = createElement(TYPES.Container)
    expect(element).toBeInstanceOf(PIXI.Container)
  })

  test('create Text', () => {
    const element = createElement(TYPES.Text, { text: 'foobar' })
    expect(element).toBeInstanceOf(PIXI.Text)
  })

  test('create Sprite', () => {
    const element = createElement(TYPES.Sprite, { texture: emptyTexture })
    expect(element).toBeInstanceOf(PIXI.Sprite)
  })

  test('create AnimatedSprite', () => {
    const element = createElement(TYPES.AnimatedSprite, { textures: [emptyTexture] })
    expect(element).toBeInstanceOf(PIXI.AnimatedSprite)
  })

  test('create ParticleContainer', () => {
    const element = createElement(TYPES.ParticleContainer)
    expect(element).toBeInstanceOf(PIXI.ParticleContainer)
  })

  test('create BitmapText', () => {
    const element = createElement(TYPES.BitmapText, { text: 'foobar', style: { font: '35px Desyrel' } })
    expect(element).toBeInstanceOf(PIXI.BitmapText)
  })

  test('create TilingSprite', () => {
    const element = createElement(TYPES.TilingSprite, { texture: emptyTexture })
    expect(element).toBeInstanceOf(PIXI.TilingSprite)
  })

  test('create Graphics', () => {
    const element = createElement(TYPES.Graphics)
    expect(element).toBeInstanceOf(PIXI.Graphics)
  })

  test('create NineSlicePlane', () => {
    const element = createElement(TYPES.NineSlicePlane, { texture: emptyTexture })
    expect(element).toBeInstanceOf(PIXI.NineSlicePlane)
  })

  test('create SimpleMesh', () => {
    const element = createElement(TYPES.SimpleMesh, { texture: emptyTexture })
    expect(element).toBeInstanceOf(PIXI.SimpleMesh)
  })

  test('create SimpleRope', () => {
    const element = createElement(
      TYPES.SimpleRope,
      {
        texture: emptyTexture,
        points: [
          new PIXI.Point(0, 0),
          new PIXI.Point(20, 20),
        ],
      },
    )
    expect(element).toBeInstanceOf(PIXI.SimpleRope)
  })

  test('get undefined', () => {
    expect(createElement('INVALID')).toBeUndefined()
  })
})

describe('element.applyProps', () => {
  let spy

  beforeAll(() => {
    spy = jest.spyOn(PIXI.Texture, 'from').mockReturnValue(emptyTexture)
  })

  afterAll(() => {
    spy.mockRestore()
  })

  test('Sprite.applyProps exists', () => {
    const element = createElement(TYPES.Sprite, { image: './image.png' })
    expect(element).toHaveProperty('applyProps')
    expect(spy).toHaveBeenCalledWith('./image.png')
  })

  test('AnimatedSprite.applyProps with images prop exists', () => {
    const element = createElement(TYPES.AnimatedSprite, { images: ['./image.png'] })
    expect(element).toHaveProperty('applyProps')
    expect(spy).lastCalledWith('./image.png')
  })

  test('AnimatedSprite.applyProps with textures prop exists', () => {
    const element = createElement(TYPES.AnimatedSprite, { textures: [PIXI.Texture.from('./image.png')] })
    expect(element).toHaveProperty('applyProps')
    expect(spy).lastCalledWith('./image.png')
  })

  test('Sprite.applyProps image', () => {
    const element = createElement(TYPES.Sprite, { image: './image.png' })
    expect(spy).lastCalledWith('./image.png')

    element.applyProps(element, { image: './image.png' }, { image: './new-image.png' })
    expect(spy).lastCalledWith('./new-image.png')
  })

  test('TilingSprite.applyProps exists', () => {
    const element = createElement(TYPES.TilingSprite, { image: './image.png' })
    expect(element).toHaveProperty('applyProps')
    expect(spy).toHaveBeenCalledWith('./image.png')
  })

  test('TilingSprite.applyProps image', () => {
    const element = createElement(TYPES.TilingSprite, { image: './image.png' })
    expect(spy).lastCalledWith('./image.png')

    element.applyProps(element, { image: './image.png' }, { image: './new-image.png' })
    expect(spy).lastCalledWith('./new-image.png')
  })

  test('SimpleRope.applyProps exists', () => {
    const element = createElement(
      TYPES.SimpleRope,
      {
        image: './image.png',
        points: [
          new PIXI.Point(0, 0),
          new PIXI.Point(20, 20),
        ],
      },
    )
    expect(element).toHaveProperty('applyProps')
    expect(spy).toHaveBeenCalledWith('./image.png')
  })

  test('SimpleRope.applyProps image', () => {
    const element = createElement(
      TYPES.SimpleRope,
      {
        image: './image.png',
        points: [
          new PIXI.Point(0, 0),
          new PIXI.Point(20, 20),
        ],
      },
    )
    expect(spy).lastCalledWith('./image.png')

    element.applyProps(
      element,
      { image: './image.png' },
      {
        image: './new-image.png',
        points: [
          new PIXI.Point(0, 0),
          new PIXI.Point(20, 20),
        ],
      },
    )
    expect(spy).lastCalledWith('./new-image.png')
  })

  test('NineSlicePlane.applyProps exists', () => {
    const element = createElement(TYPES.NineSlicePlane, { image: './image.png' })
    expect(element).toHaveProperty('applyProps')
    expect(spy).toHaveBeenCalledWith('./image.png')
  })

  test('NineSlicePlane.applyProps image', () => {
    const element = createElement(TYPES.NineSlicePlane, { image: './image.png' })
    expect(spy).lastCalledWith('./image.png')

    element.applyProps(element, { image: './image.png' }, { image: './new-image.png' })
    expect(spy).lastCalledWith('./new-image.png')
  })

  test('SimpleMesh.applyProps exists', () => {
    const element = createElement(TYPES.SimpleMesh, { image: './image.png' })
    expect(element).toHaveProperty('applyProps')
    expect(spy).toHaveBeenCalledWith('./image.png')
  })

  test('SimpleMesh.applyProps image', () => {
    const element = createElement(TYPES.SimpleMesh, { image: './image.png' })
    expect(spy).lastCalledWith('./image.png')

    element.applyProps(element, { image: './image.png' }, { image: './new-image.png' })
    expect(spy).lastCalledWith('./new-image.png')
  })

  test('Graphics.applyProps exists', () => {
    const spy = jest.fn()

    const element = createElement(TYPES.Graphics, { draw: spy })
    expect(element).toHaveProperty('applyProps')
    expect(spy).toBeCalledWith(element)
  })

  test('Graphics.applyProps draw', () => {
    const spy = jest.fn()

    const element = createElement(TYPES.Graphics, { draw: spy })
    expect(spy).toHaveBeenCalledTimes(1)

    element.applyProps(element, { draw: spy }, { draw: spy })
    expect(spy).toHaveBeenCalledTimes(2)
  })

  test('Graphics.applyProps draw prevented twice', () => {
    const draw = jest.fn()
    const props = { draw, preventRedraw: true }
    const element = createElement(TYPES.Graphics, props)
    element.applyProps(element, props, props)
    expect(draw).toHaveBeenCalledTimes(1)
  })


})

describe ('Use of cssManager', () => {
  afterEach(() => {
    jest.clearAllMocks();
  })
  test('cssManager should be defined for the element', () => {
    const element = createElement(TYPES.Container, { className: 'btn'})
    expect(element).toHaveProperty('cssManager')
    expect(cssManager).toHaveBeenCalled()
  })
  test('setProps and setCSSProps should be called', () => {
    const element = createElement(TYPES.Container, { className: 'btn', x: 0, y: 0 })
    const spy1 = jest.spyOn(element.cssManager, 'setProps')
    const spy2 = jest.spyOn(element.cssManager, 'setCSSProps')
    expect(spy1).toHaveBeenCalledWith({x: 0, y: 0});
    expect(spy2).toHaveBeenCalledWith('btn', null, {'x': 0, 'y': 0});  
  })

  test('applyProps should be called', () => {
    propsUtils.applyDefaultProps = jest.fn();
    const element = createElement(TYPES.Container, { className: 'btn', x: 0, y: 0 })
    expect(propsUtils.applyDefaultProps).toHaveBeenCalled();  
  })

})

describe('PixiComponent', () => {
  afterEach(() => {
    Object.keys(TYPES_INJECTED).forEach(k => {
      delete TYPES_INJECTED[k]
    })
  })

  test('type must be defined', () => {
    expect(() => new PixiComponent(null)).toThrow('Expect type to be defined, got `null`')
  })

  test('cannot override existing component', () => {
    expect(() => new PixiComponent('Text', {})).toThrow(
      'Component `Text` could not be created, already exists in default components.'
    )
  })

  test('inject custom component', () => {
    let lifecycle = { create: props => {} }
    new PixiComponent('Rectangle', lifecycle)
    expect(TYPES_INJECTED).toHaveProperty('Rectangle', lifecycle)
  })

  test('create injected component', () => {
    const scoped = jest.fn()

    const lifecycle = {
      create: jest.fn(() => new PIXI.Graphics()),
      didMount: jest.fn(),
      willUnmount: jest.fn(),
      applyProps: jest.fn(function() {
        scoped(this)
      }),
    }

    new PixiComponent('Rectangle', lifecycle)

    const props = { x: 100, y: 200 }
    const element = createElement('Rectangle', props)

    expect(element.didMount).toBeDefined()
    expect(element.willUnmount).toBeDefined()
    expect(element.applyProps).toBeDefined()
    expect(element).toBeInstanceOf(PIXI.Graphics)
    expect(lifecycle.create).toHaveBeenCalledTimes(1)
    expect(lifecycle.create).toHaveBeenCalledWith(props)
    expect(lifecycle.applyProps).toHaveBeenCalledTimes(1)
    expect(scoped).toHaveBeenCalledTimes(1)
    expect(scoped).toHaveBeenCalledWith(element)
  })

  test('create injected component without lifecycle methods', () => {
    new PixiComponent('Rectangle', {
      create: () => new PIXI.Graphics(),
    })

    const element = createElement('Rectangle')
    expect(element.didMount).toBeUndefined()
    expect(element.willUnmount).toBeUndefined()
    expect(element.applyProps).toBeUndefined()
  })
})
