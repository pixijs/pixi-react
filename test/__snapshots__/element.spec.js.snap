// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createElement types 1`] = `
Object {
  "BitmapText": "BitmapText",
  "Container": "Container",
  "Graphics": "Graphics",
  "Mesh": "Mesh",
  "NineSlicePlane": "NineSlicePlane",
  "ParticleContainer": "ParticleContainer",
  "Rope": "Rope",
  "Sprite": "Sprite",
  "Star": "Star",
  "Text": "Text",
  "TilingSprite": "TilingSprite",
}
`;
